src/metabase/lib_be/task/backfill_entity_ids.clj

```diff
@@ -13,150 +14,129 @@
-            (recur model row conn savepoint end-retry-count))))))
+  (->> (for [[model inner] @serdes/entity-id-cache
+             [id entity-id] inner]
+         [model id @entity-id])
+       (take *drain-batch-size*)))
 
 (defn- backfill-entity-ids!-inner
-  "Given a model, gets a batch of objects from the db and adds entity-ids"
+  "Given a model, gets a batch of objects from the db and adds entity-ids.  Returns whether there is more rows to backfill."
   [model]
   (try
-    (t2/with-transaction [^java.sql.Connection conn]
-      (let [failed-ids @failed-rows
-            id-condition (if (seq failed-ids)
-                           [:not-in failed-ids]
-                           [:!= 0])
-            rows (t2/select model :entity_id nil :id id-condition {:limit *batch-size*})]
-        (when (seq rows)
-          (log/info (str "Adding entity-ids to " (count rows) " rows of " model))
-          (doseq [{:keys [id] :as row} rows]
-            (let [savepoint (.setSavepoint conn)
-                  new-entity-id (serdes/backfill-entity-id row)
-                  failure
-                  (try
-                    (t2/update! model id {:entity_id new-entity-id})
-                    nil
-                    (catch Exception e
-                      (.rollback conn savepoint)
-                      (if (t2/select-one model :entity_id new-entity-id)
-                        (do
-                          (log/info (str "Duplicate entity-id found for " model " with id " id))
-                          (retry-insert-entity-ids! model row conn savepoint 1))
-                        e)))]
-              (when failure
-                ;; If we fail to update an individual entity id, add it to the ignore list and continue.  We'll
-                ;; retry them on next sync.
-                (add-failed-row! id)
-                (when (instance? Exception failure)
-                  (log/error (str "Exception updating entity-id for " model " with id " id))
-                  (log/error failure)))))
-          true)))
+    (if (empty? (get-rows-to-drain))
+      (let [new-rows (t2/select model :entity_id nil {:limit (if (= model :model/Field)
+                                                               ;; Backfill fields one at a time because we cache all
+                                                               ;; fields from a given table at once
+                                                               1
+                                                               *backfill-batch-size*)})]
+        (log/info "Backfill: Added " (count new-rows) " rows of " model " to the entity-id cache")
+        (seq new-rows))
+      true)
     (catch Exception e
-      ;; If we error outside of updating a single entity id, stop.  We'll retry on next sync.
-      (log/error (str "Exception updating entity-ids for " model))
+      (log/error (str "Backfill: Exception fetching entity-ids for " model))
       (log/error e))))
```

Comment:
`log/error` and `log/errorf` (and `log/warn` etc.) can take an exception as the first argument. Keeping it as one log entry is better for the logging infra IIUC.

```suggestion
      (log/error e (str "Backfill: Exception fetching entity-ids for " model)))))
```